src/librustc_lint/non_ascii_idents.rs

```diff
@@ -13,9 +14,58 @@ declare_lint! {
     "detects uncommon Unicode codepoints in identifiers"
 }
 
-declare_lint_pass!(NonAsciiIdents => [NON_ASCII_IDENTS, UNCOMMON_CODEPOINTS]);
+declare_lint! {
+    pub CONFUSABLE_IDENTS,
+    Warn,
+    "detects uncommon Unicode codepoints in identifiers"
+}
+
+declare_lint_pass!(NonAsciiIdents => [NON_ASCII_IDENTS, UNCOMMON_CODEPOINTS, CONFUSABLE_IDENTS]);
+
+fn calc_skeleton(symbol_str: &str) -> String {
+    if symbol_str == "ｓ" {
+        // FULLWIDTH LATIN SMALL LETTER S
+        return "s".to_owned();
+        // LATIN SMALL LETTER S
+    }
+    symbol_str.to_owned()
+}
 
 impl EarlyLintPass for NonAsciiIdents {
+    fn check_crate(&mut self, cx: &EarlyContext<'_>, _: &ast::Crate) {
+        use std::collections::BTreeMap;
+        let mut ascii_symbol_strs = BTreeMap::new();
+        let mut non_ascii_symbol_strs = BTreeMap::new();
+        let symbols = cx.sess.parse_sess.symbol_gallery.symbols.lock();
+        for symbol in symbols.iter() {
+            let symbol_str = symbol.as_str();
+            if symbol_str.is_ascii() {
+                ascii_symbol_strs.insert(symbol_str, *symbol);
+                continue;
+            }
+            non_ascii_symbol_strs.insert(symbol_str, *symbol);
+        }
+        drop(symbols);
+        if non_ascii_symbol_strs.is_empty() {
+            return;
+        }
+        let mut skeleton_map = FxHashMap::default();
+        for (symbol_str, symbol) in ascii_symbol_strs.iter().chain(non_ascii_symbol_strs.iter()) {
+            let skeleton = calc_skeleton(&symbol_str);
+            skeleton_map
+                .entry(skeleton)
+                .and_modify(|existing_symbol| {
+                    cx.lint(CONFUSABLE_IDENTS, |lint| {
+                        lint.build(&format!(
+                            "identifier pair considered confusable between `{}` and `{}`",
+                            existing_symbol, symbol
+                        ))
+                        .emit();
```

Comment:
Identifier spans are notably missing here, you may want to turn the "gallery" into a map and keep the spans.

Reply:
There might be multiple spans for each symbol, should i store all of them, or first occurance?

Reply:
Whatever `map.insert(symbol, span)` does seems ok.